### ISSUE TITLE
Problems with Signals and AOT on iOS (Unity 5.5.1p4)

### ISSUE THREAD

**Issue by kamend** [ssannandeji/Zenject-2019#219]
Hi,
I updated Zenject today to the latest version and I had to update all my Commands to Signals, all went well and I was able to run everything successfully in the Editor. But when I tried to run it on iOS I got the following error:

> ExecutionEngineException: Attempting to call method 'Zenject.InstanceMethodSignalHandler`2[[System.Boolean, mscorlib, Version=2.0.5.0, Culture=, PublicKeyToken=7cec85d7bea7798e],[BoolSignalHandler, Assembly-CSharp, Version=0.0.0.0, Culture=, PublicKeyToken=null]]::.ctor' for which no ahead of time (AOT) code was generated.
  at System.Reflection.MonoCMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
  at Zenject.DiContainer.InstantiateInternal (System.Type concreteType, Boolean autoInject, Zenject.InjectArgs args) [0x00000] in <filename unknown>:0 
  at Zenject.TransientProvider+<GetAllInstancesWithInjectSplit>c__Iterator0.MoveNext () [0x00000] in <filename unknown>:0 
  at Zenject.CachedProvider+<GetAllInstancesWithInjectSplit>c__Iterator0.MoveNext () [0x00000] in <filename unknown>:0 
  at Zenject.IProviderExtensions.GetAllInstances (IProvider creator, Zenject.InjectContext context, System.Collections.Generic.List`1 args) [0x00000] in <filename unknown>:0 
  at Zenject.DiContainer.SafeGetInstances (Zenject.ProviderPair providerPair, Zenject.InjectContext context) [0x00000] in <filename unknown>:0 
  at System.Linq.Enumerable+

I have omitted the full stack trace, because it is very long. So in order to test things out I made a very simple project, so I can isolate the problem. It comes from signals that have parameters and are bound to a Signal Handler.

Here is show I have setup everything. 

In my Installer I have:

```csharp
Container.Bind<SignalManager>().AsSingle();
Container.DeclareSignal<BoolSignal>();
Container.BindSignal<bool, BoolSignal>().To<BoolSignalHandler>( x => x.Run ).AsSingle();
```

My signal is declared like that:

```csharp
public class BoolSignal : Signal<bool, BoolSignal> { }
```
The handler looks like this:

```csharp
public class BoolSignalHandler  {
	public void Run(bool test) {
		Debug.Log( "Bool: "+test );
	}
}
```

Any ideas how to fix this or am I doing something wrong?

Signals without any params, works without any problems.

**Comment by kamend** [comment]
The problem seems to appear, if I use int or bool for parameters, string seems to work fine.

**Comment by dugsteen** [comment]
I am having the same problem. Any solution or workaround (besides turning bools/ints into strings) would be welcome.

**Comment by dugsteen** [comment]
A somewhat preferable workaround would be to encapsulate your int or bool parameters into a class. Classes work just fine, like strings.

**Comment by kamend** [comment]
Yeah, that will work.. I also converted my commands into services, subscribed to the Signal, so I don't change my logic. But still it will be cool, to fix this, if possible or just state that there is a problem in the documentation.

**Comment by svermeulen** [comment]
Finally had some time this weekend to take a look at this issue.  This code sums up the issue:

```csharp
public class Foo<T1>
{
    // Normally this attribute would help, but when using value types in generic arguments it does nothing useful
    [Preserve]
    public Foo()
    {
        Debug.Log("**** SUCCESS ****");
    }
}

public class Runner : MonoBehaviour
{
    static void _AotWorkaround()
    {
        // We need to uncomment this on AOT platforms for our button to work without errors
        //new Foo<int>();
    }

    public void OnGUI()
    {
        if (GUI.Button(new Rect(100, 100, 500, 100), "Attempt to Create Foo<int>"))
        {
            // Note that we have to use a generic argument here for the AOT issue to occur
            // If we instead do `CreateFoo(typeof(Foo<int>));` then we do not get
            // AOT errors (though I have no idea why)
            CreateFoo<Foo<int>>();

            // This works fine however
            //CreateFoo<Foo<string>>();
        }
    }

    void CreateFoo<T>()
    {
        // This line will fail on AOT platforms unless the above line is uncommented
        typeof(T).GetConstructors()
            .Where(x => x.GetParameters().Length == 0).Single().Invoke(new object[0]);
    }
}
```

If you create an new scene, add Runner, and then build and run on an IL2CPP platform, then click the button, you get the error.

I spent a bit of time digging into this and so far I haven't really found a workaround, other than just not using struct types.  So for now, I've changed signals to enforce this when on IL2CPP platforms.  So on IL2CPP platforms, you'll get compiler errors now due to generic constraints on the signal classes if you try to use value types.  If anyone has a better idea, I'd be happy to hear it.

**Comment by dugsteen** [comment]
You probably know all this, but from what I understand reading the Unity/iOS manuals, this is due to the fact that iOS doesn't do reflection, so Unity has to know what's going to be called _a priori_. There's no real workaround other than to create a dummy method which references Foo<int> but is never called, like you've done in _AotWorkaround. See [https://docs.unity3d.com/Manual/TroubleShootingIPhone.html](url), and look for "The game crashes with the error message “ExecutionEngineException: Attempting to JIT compile method ‘SometType`1<SomeValueType>:.ctor ()’ while running with –aot-only.”"

**Comment by svermeulen** [comment]
Fixed by #345
